util/retrieve-cluster-logs.py

```diff
@@ -0,0 +1,213 @@
+    if not args.bucket_prefix:
+        args.bucket_prefix = get_default_bucket_prefix(args.cluster)
+    args.log_group = get_log_group(args)
+    verify_bucket_exists_in_region(args)
+    return args
+
+
+def start_export_task(logs_client, args):
+    """Start the task that will export the cluster's logs to an s3 bucket, and return the task ID."""
+    LOGGER.info(
+        "Starting export of logs from log group {log_group} to s3 bucket {bucket}".format(
+            log_group=args.log_group.get("logGroupName"), bucket=args.bucket
+        )
+    )
+    start_time = datetime.fromtimestamp(args.log_group.get("creationTime") / 1000)
+    end_time = datetime.now()
+    try:
+        response = logs_client.create_export_task(
+            logGroupName=args.log_group.get("logGroupName"),
+            fromTime=int(1000 * start_time.timestamp()),
+            to=int(1000 * end_time.timestamp()),
+            destination=args.bucket,
+            destinationPrefix=args.bucket_prefix,
+        )
+    except Exception as err:
+        if "Please check if CloudWatch Logs has been granted permission to perform this operation." in str(err):
+            err_and_exit(
+                "CloudWatch Logs needs GetBucketAcl and PutObject permisson for the s3 bucket {bucket}. See {url} for "
+                "more details.".format(
+                    bucket=args.bucket,
+                    url="https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/S3ExportTasks.html#S3Permissions",
+                )
+            )
+        else:
+            err_and_exit("Unexpected error when starting export task: {}".format(err))
+    return response.get("taskId")
+
+
+def get_status_for_export_task(logs_client, task_id):
+    """Get the status for the CloudWatch export task with the given task_id."""
+    tasks = logs_client.describe_export_tasks(taskId=task_id).get("exportTasks")
+    if not tasks:
+        err_and_exit("Unable to get status for CloudWatch logs export task with ID={}".format(task_id))
+    elif len(tasks) > 2:
+        err_and_exit(
+            "More than one CloudWatch logs export task with ID={task_id}:\n{tasks}".format(
+                task_id=task_id, tasks=json.dumps(tasks, indent=2)
+            )
+        )
+    return tasks[0].get("status").get("code")
+
+
+def wait_for_task_completion(logs_client, task_id):
+    """Wait for the CloudWatch logs export task given by task_id to finish."""
+    LOGGER.info("Waiting for export task with task ID {} to finish.".format(task_id))
+    status = "PENDING"
+    still_running_statuses = ("PENDING", "PENDING_CANCEL", "RUNNING")
+    while status in still_running_statuses:
+        status = get_status_for_export_task(logs_client, task_id)
+    return status
+
+
+def export_logs_to_s3(args):
+    """Export the contents of a cluster's CloudWatch log group to an s3 bucket."""
+    logs_client = boto3.client("logs", region_name=args.region)
+    task_id = start_export_task(logs_client, args)
```

Comment:
Shall we also check that there isn't a running export before starting a new one?

Reply:
That is a very good idea, depending on how we want to handle that situation. I need to confirm this, but my understanding is that the logs client's `create_export_task` queues the export task if there's already one running in that region.

My initial reaction would be to not explicitly check for an ongoing task, but instead check if the newly created export task is queued based on the export task's status code. Then if it is, log that information and any information made available by the API regarding where in the queue the task is (which is something else I'd need to look into).

The alternative would be to check for ongoing export tasks and then fail out of the script if one exists.

Your thoughts?

Reply:
I'd say if we decide to queue additional `create_export_task`s then let's make sure the script waits for it to be completed doesn't fail because there are multiple tasks in the queue. Alternatively we could just simplify the approach for now and fail fast if a task is already in the queue.

A future enhancement might be to also SIGINT and perform a cleanhup when this occurs. But we can leave this for later.

Reply:
The script currently waits for the task to start and complete, so I'll leave it as is for now unless you feel strongly about changing it. I added a comment at the top of the script with a TODO for catching the SIGINT and cleaning up appropriately.